keps/draft-20180806-namespace-initializer.md

```diff
@@ -0,0 +1,345 @@
+
+### Goals
+
+- Core Kubernetes API changes (extend Namespace resource with complete backward compatibility) for namespace initialization
+- Extendability on namespace initializers
+- Consideration on scalability
+- Recommendations for client/CLI changes (no change required to keep existing clients/CLI working, but change for better)
+
+### Non-Goals
+
+- The details of existing Initializers mechanism
+- Namespace initializers configured to be applied to a subset of namespaces
+- Application specific use of namespace initialization
+
+## Proposal
+
+### User Stories
+
+#### Automatic Policy Population
+
+In a large enterprise, a single Kubernetes cluster is shared by multiple engineering teams.
+The cluster administrators allow the engineering teams to do self-service namespace creation (using `kubectl create namespace`) to make the process simpler,
+but they still want to enforce the policies in each namespace automatically, including RBAC, ResourceQuota, NetworkPolicy etc.
+They define these policies in a centralized places (e.g. git, corp AD, etc.),
+and use a controller watching namespace creation and populate the predefined policies during namespace creation before the creator gains full permission inside the namespace.
+
+This is the user story demonstrated by the [Namespace Population](https://github.com/kubernetes/community/pull/2177) proposal.
+
+#### Automated External Service Provisioning
+
+The Kubernetes cluster administrator wants to associate namespaces with log analysis system backed by ElasticSearch.
+During the namespace creation, they want to automatically provision ElasticSearch with a new index dedicated for logs sent from this namespace.
+They deploy a controller to perform the provisioning when watch a new namespace creation,
+and they don't want the users to access the namespace before the provisioning is completed.
+
+### Implementation Details
+
+The mechanism is derived from exising [Initializers](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#initializers).
+For more background and initial discussion, please read comments in [Namespace Population KEP](https://github.com/kubernetes/community/pull/2177).
+
+Briefly, namespace initialization involves a few components:
+
+- A built-in admission controller `NamespaceInitializers`
+- A resource of type `NamespaceInitializerConfiguration`
+- Some logic in storage layer to keep the backward-compatible behavior of the API
+- A few namespace initialization controllers: custom controllers which initialize a namespace by creating resources inside the namespace
+
+They work in the process of:
+
+- The namespace initialization controllers register themselves by creating `NamespaceInitializerConfiguration` resources
+- When a namespace is newly created
+  - The `NamespaceInitializers` admission controller aggregates all `NamespaceInitializersConfiguration` resources and generate `spec.Initializers` in this namespace resource
```

Comment:
how do namespaces decide whether they are subject to being run through an admission plugin w.r.t. `NamespaceSelector`?

it gets a little weird to use this to hook in policy when the thing used to select whether it applies to a namespace is an attribute of the incoming object:

https://github.com/kubernetes/kubernetes/blob/daf17e2aee3c796b4e741d8d21ad20382237dd6d/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/namespace/matcher.go#L89-L98

Reply:
IIRC namespace objects have their own labels tested before running through the admission controller, for better or worse.

Reply:
@deads2k Yes. We can definitely tell people to use a webhook admission plugin. However, adding a webhook admission plugin requires additional work, including setting up a webserver with TLS certificates, creating webhook configurations, exposing services, etc. And the stability of webhook plugins also impacts the request in API path. With `NamespaceInitializersConfiguration` saves these extra work that people only need to write a controller. It also helps we can focus on the stability of NamespaceInitializers admin controller.

This design is completely copied from existing Initializers: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#initializers, and IMO the original design is good as it offers a simplified configuration in additional to writing own admission webhook plugin.

Reply:
@liggitt Not sure if I understand the comment correctly. The `NamespaceInitializers` admission controller runs through all namespaces without filtering. And currently there's no filtering/selection mechanism in `NamespaceInitializerConfiguration`. That means, they are applied to all newly created namespaces, and leave to the actual controllers performing initialization work to decide what to do with a particular namespace.

Reply:
For the primary use case given (self-service namespace creation/population - https://github.com/kubernetes/community/pull/2177), a webhook is still required to capture the info about the creating user, right? the initializer name popualted in the namespace doesn't have enough information for auto-policy-grants to know who to grant permissions to

Reply:
Yes, for the particular use case in #2177. I'd love to separate this problem from this PR and discuss that in #2177. However, one of the major dependencies on this PR is that, we need this initialization mechanism to make sure the compatibility of API behavior, that an existing CLI/script won't break when namespace is being initialized. And that's also the reason we need this change in core Kubernetes. There's no other way to achieve that, e.g. a webhook admission plugin.

Reply:
> @deads2k Yes. We can definitely tell people to use a webhook admission plugin. However, adding a webhook admission plugin requires additional work, including setting up a webserver with TLS certificates, creating webhook configurations, exposing services, etc. And the stability of webhook plugins also impacts the request in API path. With `NamespaceInitializersConfiguration` saves these extra work that people only need to write a controller. It also helps we can focus on the stability of NamespaceInitializers admin controller.

This design is trying to avoid that at the cost of adding additional complexity to core.  In addition, doing it this way restricts options that admission plugins would have about precisely when and how particular initializers are applied.  Using a webhook, I'm sure that the openshift use-cases could be covered.  Using what's described here, I'm not guaranteed that it would be.

Reply:
> about precisely when and how particular initializers are applied

Is that specifically talking about the order of initializers? If yes, using Webhook admission plugin (I mean [MutatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook)) doesn't guarantee the order. Actually they are the same: creating `MutatingWebhookConfiguration` vs creating `NamespaceInitializerConfiguration`. And it's really difficult to define/enforce the ordering unless the webhook plugins (or initializers) are clearly predefined during cluster bootstrapping time and disallow dynamic deployment of arbitrary webhooks/initializers.

Anyway, with `NamespaceInitializerConfiguration`, it doesn't block the use of webhook admission plugin. It's still possible/reasonable to use webhook admission plugin depending on a particular use case. However we did hear the complain about the extra work/complexity to build a webhook while the developer only want a controller to populate a namespace when it's being initialized. That demonstrates the value of `NamespaceInitializerConfiguration`.

Alternatively, I'm open to not implementing `NamespaceInitializerConfiguration` at the beginning, and only focus on initialization mechanism, and with that alone, webhook plugin is required to add initializers.

Reply:
Maybe we want to have an easy to consume initializer impl as part of controller-runtime that someone can quickly iterate on.